### ISSUE TITLE
Save 75% of platform message size 🤦🏻‍♂️

### ISSUE THREAD

**Issue by zubairov** [elasticio/sailor-nodejs#80]
This [place here](https://github.com/elasticio/sailor-nodejs/blob/master/lib/cipher.js#L30) returns a string with **base64** encoded string in it. Base64 binary is roughly [75% larger than binary data](https://stackoverflow.com/questions/4715415/base64-what-is-the-worst-possible-increase-in-space-usage). The [`publish` method](http://www.squaremobius.net/amqp.node/channel_api.html#channel_publish)  that sends data to the queue accepts a binary `Buffer` so it makes little sense to send a based64 string over it.
BTW `publish` method copies the data, so data copied 3 times - [here](https://github.com/elasticio/sailor-nodejs/blob/master/lib/cipher.js#L30), [here when new Buffer is created](https://github.com/elasticio/sailor-nodejs/blob/master/lib/amqp.js#L107)  and [here when data is published to the queue](https://github.com/elasticio/sailor-nodejs/blob/master/lib/amqp.js#L107)

**Comment by jhorbulyk** [comment]
> The main problem here when fixing it - maintain backwards compatibility.

All sailor versions need to be updated at the same time. :(

**Comment by ghaiklor** [comment]
@zubairov either I did not understand the problem, but I think we can make it backward compatible. It is a little bit stinky, but should work.

So, if you want to have a possibility to exchange both base64 or Buffer, you can predict the format of the message you got.

Pseudo-code

```js
function decryptIV(encData, options) {
    const isBase64 = Buffer.from(encData, 'base64').toString('base64') === encData;
    const inputFormat = isBase64 ? 'base64' : 'buffer';
    const result = cipher.update(encData, inputFormat, 'utf-8') + cipher.final('utf-8');

    return result;
}
```

The idea behind this trick is simple:

1) You got a message from rabbit (it can be both base64 or buffer, you don't know for sure)
2) You are trying to decode message as base64 -> encode back to base64 and *IF* encoded data equals to encoded messages - we can be sure that is base64
3) Otherwise, it message was not base64, it can not be decoded/encoded properly and *definitely* will not be equal to the source message. In that case, we are assuming that message was sent in Buffer format.

Now, regard to compatibility:

New sailor versions in `encrypt` would use Buffer format, so all the messages will be sent in Buffer format.
Plus, `decrypt` method will have a mechanism to determine the format of message and handle appropriate (as described above).

Old sailor versions in `encrypt` would use base64 format and `decrypt` would use base64 as well.

Another case, when message is sent from old to new sailor:

Old `encrypt` method will send base64 format, but since our new sailor is capable of determine that it is base64 it will just decrypt it as before. Otherwise, it will take it as a buffer if message sent from new sailor.

Hope I'm right about the problem you saying

**Comment by zubairov** [comment]
@ghaiklor ok, what happens when "new" version send a message in `Buffer` format to "old" sailor?

**Comment by ghaiklor** [comment]
@zubairov dope, I didn't think about this one, shame on me 😆 

We _could_ ignore this issue for now and just increase the limits for pods where sailor is running, but... it it a short-term solution.

**Comment by zubairov** [comment]
It's not really a `pod` related but RabbitMQ related issue :) but yes, we need to build a message versioning support into `sailor` ASAP to make sure we are protected from such issues in future.

**Comment by ghaiklor** [comment]
RabbitMQ has no issues with handling big messages. What do you mean by that ?

**Comment by zubairov** [comment]
Yes, correct, my statement should be - we produce unnecessary load on RabbitMQ due to sailor implementation details

**Comment by ghaiklor** [comment]
I'm going to think about any other possible solutions for problem with huge messages, but for now we have the following facts:

- Most of the steps have been killed because of Out Of Memory
- seems like OOM is produced by sailor due to dirty implementation (_do we have the same problem with Java sailor?_)
- it has nothing to do with RabbitMQ, since its amqp protocol is based on tcp and all the "split-send" stuff is happening under the hood

For now, I do not see any ways to solve it without touching sailor (_expect increasing memory limits_).

**Comment by zubairov** [comment]
@ghaiklor I don't think it's the solution for huge messages, only a minor efficiency gain, not more than that.
